e2e/tests/proxy_test.go

```diff
@@ -287,3 +291,32 @@ func TestProxyUpstream(t *testing.T) {
 		t.Fatalf("%s via header not found", forwarder.UpstreamProxyServiceName)
 	}
 }
+
+func TestProxyReuseConnection(t *testing.T) {
+	c := newClient(t, "", func(tr *http.Transport) {
+		var (
+			d      net.Dialer
+			dialed atomic.Bool
+		)
+		tr.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
+			if dialed.CompareAndSwap(false, true) {
+				return d.DialContext(ctx, network, addr)
+			}
+			return nil, errors.New("only one dial is allowed")
+		}
+	})
+
+	for _, h := range []string{
+		"http://wronghost1",
+		"http://wronghost2",
+	} {
+		res := c.GET(h, func(req *http.Request) {
+			var body bytes.Buffer
```

Comment:
What if you try to use io.Pipe for a body or a limit reader of 10MB reading zeros?

Reply:
Added io.Pipe with zeros to be read.